### PR TITLE
Use the engine interface with AFL on a trial basis.

### PR DESCRIPTION
**Pull request by mbarbella-chromium**
Not yet ready for review. This still needs a bit of cleanup and much more careful testing.

### CODE REVIEW

**Comment by mbarbella-chromium** [comment]
Still needs a small amount of work to get tests passing. Feel free to take a look now if you want, but I'll update this again once everything is green.

**Comment by mbarbella-chromium** [comment]
Tests are now passing.

File: src/python/bot/fuzzers/afl/engine.py
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:28]: nit: MINIJAIL is now completely disabled, so can skip this.

File: src/python/bot/fuzzers/afl/engine.py
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:138]: this doesn't do anything in the reproduce case right?

File: src/python/bot/fuzzers/afl/engine.py
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:83]: rather than set an env var here to be used later, can we just modify runner's fuzz method to accept this argument, which is then passed to afl's get_fuzz_timeout?

File: src/python/bot/fuzzers/afl/engine.py
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:104]: nit: log_error to match libFuzzer: https://github.com/google/clusterfuzz/blob/0b5c023eca9e3aac41faba17da8f341c0ca2ddc7/src/python/bot/fuzzers/libFuzzer/engine.py#L283

File: src/python/bot/tasks/fuzz_task.py
Comment (oliverchang) [inline src/python/bot/tasks/fuzz_task.py:1912]: I'm guessing the plan is to turn this on for reproduction too once we verify that fuzzing seems to work?

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py
Comment (oliverchang) [inline src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:83]: Can we move/copy some more tests from afl_launcher_integration_test here which tests that we are actually catching crashes properly?

**Review by oliverchang — COMMENTED** [review]
nice! some initial comments.

File: src/python/bot/fuzzers/afl/engine.py
Comment (jonathanmetzman) [reply src/python/bot/fuzzers/afl/engine.py:28]: Really? What replaced it for misbehaved OSS-Fuzz targets like rar?

File: src/python/bot/fuzzers/afl/engine.py
Comment (jonathanmetzman) [reply src/python/bot/fuzzers/afl/engine.py:83]: +1

File: src/python/bot/fuzzers/afl/engine.py
Comment (jonathanmetzman) [reply src/python/bot/fuzzers/afl/engine.py:138]: +1

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py
Comment (jonathanmetzman) [inline src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:71]: nit: Don't even assign if we aren't using the result.

File: src/python/bot/fuzzers/afl/engine.py
Comment (oliverchang) [reply src/python/bot/fuzzers/afl/engine.py:28]: No replacement. @inferno-chromium disabled it and was monitoring.

File: src/python/bot/fuzzers/afl/engine.py
Comment (mbarbella-chromium) [reply src/python/bot/fuzzers/afl/engine.py]: Done!

File: src/python/bot/fuzzers/afl/engine.py
Comment (mbarbella-chromium) [reply src/python/bot/fuzzers/afl/engine.py]: Right, removed.

File: src/python/bot/tasks/fuzz_task.py
Comment (mbarbella-chromium) [reply src/python/bot/tasks/fuzz_task.py:1912]: We should actually do this now. Made that change as well. It'll be gated behind a separate job type initially, so there's no harm in enabling in both cases.

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py
Comment (mbarbella-chromium) [reply src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py]: Done!

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py
Comment (mbarbella-chromium) [reply src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:83]: Sure, I've added one for testing a crash case. That seems to be failing, but I'm working on a fix for that now. Just publishing these comments in the meantime.

For the others, whenever we stop using the launcher we should make sure all of those tests are preserved. This is essentially just calling into the launcher, so I'm satisfied that so long as we know that we can handle the basic cases (adding stuff to the corpus, processing crashes, reproducing) that the corner cases tested there will function properly.

Comment (mbarbella-chromium) [reply src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:86]: Added another new test, and both are now passing.

File: src/python/bot/testcase_manager.py
Comment (oliverchang) [inline src/python/bot/testcase_manager.py:584]: Can just put this in engine.get() instead of duplicating the checks in two places.

File: src/python/bot/tasks/fuzz_task.py
Comment (oliverchang) [reply src/python/bot/tasks/fuzz_task.py:1920]: Ok, we need to be very careful here to make sure existing testcases don't break. Please verify this manually on a number of testcases after landing (prior to our progression cron).

File: src/python/bot/tasks/fuzz_task.py
Comment (oliverchang) [inline src/python/bot/tasks/fuzz_task.py:1909]: What is this for?

**Review by oliverchang — APPROVED** [review]
Thanks. Some additional comments, but LGTM overall. Please test this carefully.

File: src/python/bot/tasks/fuzz_task.py
Comment (mbarbella-chromium) [reply src/python/bot/tasks/fuzz_task.py]: This was incorrectly included when resolving a conflict. Removed.

File: src/python/bot/testcase_manager.py
Comment (mbarbella-chromium) [reply src/python/bot/testcase_manager.py]: Good call. Done.

**Comment by mbarbella-chromium** [comment]
I'll hold off on landing this until after the progression cron runs to give myself a bit more time for manual testing. All comments should be addressed, and tests are all passing.

**Comment by mbarbella-chromium** [comment]
Spoke to Oliver via chat. We'll just land this now and monitor.

**Pull request merged**
